File: src/indexer/adapter/csharp.ts
import Parser from "tree-sitter";
import type { Tree } from "tree-sitter";
import type { LanguageAdapter } from "./LanguageAdapter.js";
import {
  getParser,
  clearCache as clearGrammarCache,
} from "../treesitter/grammarLoader.js";
import type {
  ExtractedSymbol,
  ExtractedCall,
} from "../treesitter/extractCalls.js";
import type { ExtractedImport } from "../treesitter/extractImports.js";
import { findEnclosingSymbol as findEnclosingSymbolUtil } from "../treesitter/symbolUtils.js";

class CSharpAdapter implements LanguageAdapter {
  languageId = "csharp";
  fileExtensions = [".cs"] as const;

  private parser: Parser | null = null;

  getParser(): Parser | null {
    if (!this.parser) {
      this.parser = getParser("csharp");
    }
    return this.parser;
  }

  parse(content: string, filePath: string): Tree | null {
    const parser = this.getParser();
    if (!parser) {
      return null;
    }

    try {
      // Use 1MB buffer to handle files >32KB (tree-sitter default limit)
      const tree = parser.parse(content, undefined, {
        bufferSize: 1024 * 1024,
      });

      if (!tree || tree.rootNode.hasError) {
        return null;
      }

      return tree;
    } catch (error) {
      process.stderr.write(
        `[sdl-mcp] Failed to parse C# file ${filePath}: ${error instanceof Error ? error.message : String(error)}\n`,
      );
      return null;
    }
  }

  extractSymbols(
    tree: Tree,
    _content: string,
    filePath: string,
  ): ExtractedSymbol[] {
    const symbols: ExtractedSymbol[] = [];
    const normalizedPath = filePath.replace(/\\/g, "/");
    this.traverseASTForSymbols(tree.rootNode, symbols, normalizedPath);
    return symbols;
  }

  extractImports(
    tree: Tree,
    _content: string,
    _filePath: string,
  ): ExtractedImport[] {
    return this.extractImportsImpl(tree);
  }

  extractCalls(
    tree: Tree,
    _content: string,
    _filePath: string,
    extractedSymbols: ExtractedSymbol[],
  ): ExtractedCall[] {
    return this.extractCallsImpl(tree, extractedSymbols);
  }

  private traverseASTForSymbols(
    node: Parser.SyntaxNode,
    symbols: ExtractedSymbol[],
    filePath: string,
  ): void {
    switch (node.type) {
      case "namespace_declaration":
        const namespaceSymbol = this.processNamespaceDeclaration(
          node,
          filePath,
        );
        if (namespaceSymbol) symbols.push(namespaceSymbol);
        break;

      case "class_declaration":
        const classSymbol = this.processClassDeclaration(node, filePath);
        if (classSymbol) symbols.push(classSymbol);
        break;

      case "interface_declaration":
        const interfaceSymbol = this.processInterfaceDeclaration(
          node,
          filePath,
        );
        if (interfaceSymbol) symbols.push(interfaceSymbol);
        break;

      case "struct_declaration":
        const structSymbol = this.processStructDeclaration(node, filePath);
        if (structSymbol) symbols.push(structSymbol);
        break;

      case "enum_declaration":
        const enumSymbol = this.processEnumDeclaration(node, filePath);
        if (enumSymbol) symbols.push(enumSymbol);
        break;

      case "record_declaration":
        const recordSymbol = this.processRecordDeclaration(node, filePath);
        if (recordSymbol) symbols.push(recordSymbol);
        break;

      case "method_declaration":
        const methodSymbol = this.processMethodDeclaration(node, filePath);
        if (methodSymbol) symbols.push(methodSymbol);
        break;

      case "constructor_declaration":
        const constructorSymbol = this.processConstructorDeclaration(
          node,
          filePath,
        );
        if (constructorSymbol) symbols.push(constructorSymbol);
        break;

      case "property_declaration":
        const propertySymbol = this.processPropertyDeclaration(node, filePath);
        if (propertySymbol) symbols.push(propertySymbol);
        break;

      case "field_declaration":
        const fieldSymbol = this.processFieldDeclaration(node, filePath);
        if (fieldSymbol) symbols.push(fieldSymbol);
        break;
    }

    for (const child of node.children) {
      this.traverseASTForSymbols(child, symbols, filePath);
    }
  }

  private extractIdentifier(node: Parser.SyntaxNode | null): string | null {
    if (!node) return null;

    if (node.type === "identifier") {
      return node.text;
    }

    for (const child of node.children) {
      if (child.type === "identifier") {
        return child.text;
      }
    }

    return null;
  }

  private extractRange(node: Parser.SyntaxNode) {
    const start = node.startPosition;
    const end = node.endPosition;

    return {
      startLine: start.row + 1,
      startCol: start.column,
      endLine: end.row + 1,
      endCol: end.column,
    };
  }

  private extractVisibility(
    node: Parser.SyntaxNode,
  ): "public" | "private" | "protected" | "internal" | undefined {
    for (const child of node.children) {
      if (child.type === "modifiers") {
        for (const modifier of child.children) {
          if (modifier.type === "accessibility_modifier") {
            const modifierText = modifier.text;
            if (modifierText === "public") return "public";
            if (modifierText === "private") return "private";
            if (modifierText === "protected") return "protected";
            if (modifierText === "internal") return "internal";
          }
        }
      }
    }
    return undefined;
  }

  private extractModifiers(node: Parser.SyntaxNode): string[] {
    const modifiers: string[] = [];
    const modifiersNode = node.children.find((c) => c.type === "modifiers");
    if (modifiersNode) {
      for (const child of modifiersNode.children) {
        if (child.type === "accessibility_modifier") {
          modifiers.push(child.text);
        } else if (
          child.type === "static_modifier" ||
          child.type === "async_modifier" ||
          child.type === "readonly_modifier" ||
          child.type === "override_modifier" ||
          child.type === "virtual_modifier" ||
          child.type === "abstract_modifier" ||
          child.type === "sealed_modifier" ||
          child.type === "unsafe_modifier"
        ) {
          modifiers.push(child.text);
        }
      }
    }
    return modifiers;
  }

  private extractTypeParameters(node: Parser.SyntaxNode): string[] {
    const generics: string[] = [];
    const typeParams = node.children.find(
      (c) => c.type === "type_parameter_list",
    );
    if (typeParams) {
      for (const child of typeParams.children) {
        if (child.type === "type_parameter") {
          generics.push(child.text);
        }
      }
    }
    return generics;
  }

  private extractParameters(
    node: Parser.SyntaxNode,
  ): Array<{ name: string; type?: string }> {
    const params: Array<{ name: string; type?: string }> = [];
    const paramList = node.children.find((c) => c.type === "parameter_list");
    if (paramList) {
      for (const child of paramList.children) {
        if (child.type === "parameter") {
          const identifier = this.extractIdentifier(child);
          const typeAnnotation = child.children.find((c) => c.type === "type");
          if (identifier) {
            params.push({
              name: identifier,
              type: typeAnnotation ? typeAnnotation.text : undefined,
            });
          }
        }
      }
    }
    return params;
  }

  private extractReturnType(node: Parser.SyntaxNode): string | undefined {
    const typeNode = node.children.find((c) => c.type === "type");
    if (typeNode) {
      return typeNode.text;
    }
    return undefined;
  }

  private processNamespaceDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = node.childForFieldName("name");
    if (!name) return null;

    return {
      nodeId: `${filePath}:${name.text}:0`,
      kind: "module",
      name: name.text,
      exported: true,
      range: this.extractRange(node),
    };
  }

  private processClassDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);
    const generics = this.extractTypeParameters(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "class",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
      signature: {
        params: [],
        generics: generics.length > 0 ? generics : undefined,
      },
    };
  }

  private processInterfaceDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);
    const generics = this.extractTypeParameters(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "interface",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
      signature: {
        params: [],
        generics: generics.length > 0 ? generics : undefined,
      },
    };
  }

  private processStructDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);
    const generics = this.extractTypeParameters(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "class",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
      signature: {
        params: [],
        generics: generics.length > 0 ? generics : undefined,
      },
    };
  }

  private processEnumDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "type",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
    };
  }

  private processRecordDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);
    const generics = this.extractTypeParameters(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "class",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
      signature: {
        params: [],
        generics: generics.length > 0 ? generics : undefined,
      },
    };
  }

  private processMethodDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);
    const modifiers = this.extractModifiers(node);
    const params = this.extractParameters(node);
    const returns = this.extractReturnType(node);

    const isAsync = modifiers.includes("async");

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "method",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
      signature: {
        params,
        returns: returns || (isAsync ? "Task" : "void"),
      },
    };
  }

  private processConstructorDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);
    const params = this.extractParameters(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "constructor",
      name: "constructor",
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
      signature: {
        params,
      },
    };
  }

  private processPropertyDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "variable",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
    };
  }

  private processFieldDeclaration(
    node: Parser.SyntaxNode,
    filePath: string,
  ): ExtractedSymbol | null {
    const name = this.extractIdentifier(node);
    if (!name) return null;

    const visibility = this.extractVisibility(node);

    return {
      nodeId: `${filePath}:${name}:${this.hashCode(node.id)}`,
      kind: "variable",
      name,
      exported: visibility === "public",
      visibility,
      range: this.extractRange(node),
    };
  }

  private hashCode(id: number): number {
    return Math.abs(id);
  }

  private extractImportsImpl(tree: Tree): ExtractedImport[] {
    const imports: ExtractedImport[] = [];
    this.traverseForImports(tree.rootNode, imports);
    return imports;
  }

  private traverseForImports(
    node: Parser.SyntaxNode,
    imports: ExtractedImport[],
  ): void {
    if (node.type === "using_directive") {
      this.processUsingDirective(node, imports, false);
    } else if (node.type === "global_using_directive") {
      this.processUsingDirective(node, imports, true);
    }

    for (const child of node.children) {
      this.traverseForImports(child, imports);
    }
  }

  private processUsingDirective(
    node: Parser.SyntaxNode,
    imports: ExtractedImport[],
    _isGlobal: boolean,
  ): void {
    let isStatic = false;
    let alias: string | undefined;
    let namespace = "";
    let aliasIdentifier = "";

    for (const child of node.children) {
      if (child.type === "static_modifier") {
        isStatic = true;
      } else if (child.type === "identifier") {
        if (!aliasIdentifier) {
          aliasIdentifier = child.text;
        } else {
          namespace = child.text;
        }
      } else if (child.type === "type" || child.type === "qualified_name") {
        const name = this.extractQualifiedName(child);
        if (name) {
          if (!namespace) {
            namespace = name;
          } else {
            namespace = name;
          }
        }
      } else if (child.type === "name_equals") {
        const identifier = child.children.find((c) => c.type === "identifier");
        if (identifier) {
          alias = identifier.text;
        }
      }
    }

    if (!namespace && aliasIdentifier) {
      namespace = aliasIdentifier;
    }

    if (namespace) {
      const isRelative = false;
      const isExternal =
        !namespace.startsWith("System") && !namespace.startsWith("global::");

      const importData: ExtractedImport = {
        specifier: namespace,
        isRelative,
        isExternal,
        imports: isStatic ? ["*"] : [],
        isReExport: false,
      };

      if (alias) {
        importData.defaultImport = alias;
      }

      imports.push(importData);
    }
  }

  private extractQualifiedName(node: Parser.SyntaxNode): string | null {
    if (node.type === "identifier") {
      return node.text;
    }

    if (node.type === "qualified_name") {
      const parts: string[] = [];
      for (const child of node.children) {
        if (child.type === "identifier" || child.type === "qualified_name") {
          const name = this.extractQualifiedName(child);
          if (name) parts.push(name);
        }
      }
      return parts.length > 0 ? parts.join(".") : null;
    }

    return null;
  }

  private extractCallsImpl(
    tree: Tree,
    extractedSymbols: ExtractedSymbol[],
  ): ExtractedCall[] {
    const calls: ExtractedCall[] = [];
    const seenCallNodes = new Set<number>();

    const symbolMap = new Map<string, ExtractedSymbol>();
    for (const symbol of extractedSymbols) {
      symbolMap.set(symbol.name, symbol);
    }

    this.traverseForCalls(
      tree.rootNode,
      calls,
      seenCallNodes,
      symbolMap,
      extractedSymbols,
    );

    return calls;
  }

  private traverseForCalls(
    node: Parser.SyntaxNode,
    calls: ExtractedCall[],
    seenCallNodes: Set<number>,
    symbolMap: Map<string, ExtractedSymbol>,
    extractedSymbols: ExtractedSymbol[],
  ): void {
    if (node.type === "invocation_expression") {
      this.processInvocation(
        node,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    } else if (node.type === "object_creation_expression") {
      this.processObjectCreation(
        node,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    } else if (node.type === "await_expression") {
      this.processAwaitExpression(
        node,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    }

    for (const child of node.children) {
      this.traverseForCalls(
        child,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    }
  }

  private processInvocation(
    node: Parser.SyntaxNode,
    calls: ExtractedCall[],
    seenCallNodes: Set<number>,
    symbolMap: Map<string, ExtractedSymbol>,
    extractedSymbols: ExtractedSymbol[],
  ): void {
    const nodeId = node.id;
    if (seenCallNodes.has(nodeId)) return;
    seenCallNodes.add(nodeId);

    const funcNode = node.childForFieldName("function");
    if (!funcNode) return;

    let calleeIdentifier = "";
    let callType: ExtractedCall["callType"] = "function";
    let isResolved = false;
    let calleeSymbolId: string | undefined;

    if (funcNode.type === "identifier") {
      calleeIdentifier = funcNode.text;
      const symbol = symbolMap.get(calleeIdentifier);
      if (symbol) {
        isResolved = true;
        calleeSymbolId = symbol.nodeId;
      }
    } else if (funcNode.type === "member_access_expression") {
      callType = "method";
      const obj =
        funcNode.childForFieldName("expression") ||
        funcNode.children.find((c) => c.type === "identifier");
      const prop =
        funcNode.childForFieldName("name") ||
        funcNode.children.find((c) => c.type === "identifier" && c !== obj);
      if (obj && prop && obj !== prop) {
        calleeIdentifier = `${obj.text}.${prop.text}`;
        if (obj.text === "this" || obj.text === "base") {
          isResolved = symbolMap.has(prop.text);
          if (isResolved) {
            const resolvedSymbol = symbolMap.get(prop.text);
            if (resolvedSymbol) {
              calleeSymbolId = resolvedSymbol.nodeId;
            }
          }
        }
      }
    } else if (funcNode.type === "generic_name") {
      const identifier = funcNode.childForFieldName("name");
      if (identifier) {
        calleeIdentifier = identifier.text;
        const symbol = symbolMap.get(calleeIdentifier);
        if (symbol) {
          isResolved = true;
          calleeSymbolId = symbol.nodeId;
        }
      }
    }

    const range = {
      startLine: node.startPosition.row + 1,
      startCol: node.startPosition.column,
      endLine: node.endPosition.row + 1,
      endCol: node.endPosition.column,
    };

    const callerNodeId = this.findEnclosingSymbol(node, extractedSymbols);

    calls.push({
      callerNodeId,
      calleeIdentifier,
      isResolved,
      callType,
      calleeSymbolId,
      range,
    });

    this.extractChainedCalls(
      funcNode,
      callerNodeId,
      calls,
      seenCallNodes,
      symbolMap,
      extractedSymbols,
    );
  }

  private processObjectCreation(
    node: Parser.SyntaxNode,
    calls: ExtractedCall[],
    seenCallNodes: Set<number>,
    symbolMap: Map<string, ExtractedSymbol>,
    extractedSymbols: ExtractedSymbol[],
  ): void {
    const nodeId = node.id;
    if (seenCallNodes.has(nodeId)) return;
    seenCallNodes.add(nodeId);

    const typeNode = node.childForFieldName("type");
    if (!typeNode) return;

    let calleeIdentifier = "";
    let isResolved = false;
    let calleeSymbolId: string | undefined;

    if (typeNode.type === "identifier") {
      calleeIdentifier = `new ${typeNode.text}`;
      const symbol = symbolMap.get(typeNode.text);
      if (symbol) {
        isResolved = true;
        calleeSymbolId = symbol.nodeId;
      }
    } else if (typeNode.type === "generic_name") {
      const identifier = typeNode.childForFieldName("name");
      if (identifier) {
        calleeIdentifier = `new ${identifier.text}`;
        const symbol = symbolMap.get(identifier.text);
        if (symbol) {
          isResolved = true;
          calleeSymbolId = symbol.nodeId;
        }
      }
    }

    const range = {
      startLine: node.startPosition.row + 1,
      startCol: node.startPosition.column,
      endLine: node.endPosition.row + 1,
      endCol: node.endPosition.column,
    };

    const callerNodeId = this.findEnclosingSymbol(node, extractedSymbols);

    calls.push({
      callerNodeId,
      calleeIdentifier,
      isResolved,
      callType: "constructor",
      calleeSymbolId,
      range,
    });
  }

  private processAwaitExpression(
    node: Parser.SyntaxNode,
    calls: ExtractedCall[],
    seenCallNodes: Set<number>,
    symbolMap: Map<string, ExtractedSymbol>,
    extractedSymbols: ExtractedSymbol[],
  ): void {
    const expression = node.firstChild;
    if (!expression) return;

    if (expression.type === "invocation_expression") {
      this.processInvocation(
        expression,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    } else if (expression.type === "object_creation_expression") {
      this.processObjectCreation(
        expression,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    }
  }

  private extractChainedCalls(
    node: Parser.SyntaxNode,
    callerNodeId: string,
    calls: ExtractedCall[],
    seenCallNodes: Set<number>,
    symbolMap: Map<string, ExtractedSymbol>,
    extractedSymbols: ExtractedSymbol[],
  ): void {
    if (!node) return;

    if (node.type === "member_access_expression") {
      const obj =
        node.childForFieldName("expression") ||
        node.children.find((c) => c.type === "identifier");
      if (obj?.type === "invocation_expression") {
        const callNode = obj;
        if (!seenCallNodes.has(callNode.id)) {
          this.processInvocation(
            callNode,
            calls,
            seenCallNodes,
            symbolMap,
            extractedSymbols,
          );
        }
      }
    }

    for (const child of node.children) {
      this.extractChainedCalls(
        child,
        callerNodeId,
        calls,
        seenCallNodes,
        symbolMap,
        extractedSymbols,
      );
    }
  }

  private findEnclosingSymbol(
    node: Parser.SyntaxNode,
    symbols: ExtractedSymbol[],
  ): string {
    return findEnclosingSymbolUtil(node, symbols);
  }
}

function clearCache(): void {
  clearGrammarCache("csharp");
}

export { CSharpAdapter, clearCache };

File: native/src/extract/symbols/csharp.rs
use tree_sitter::Node;

use crate::types::NativeParsedSymbol;

use super::common::{find_child_by_kind, find_child_node, make_symbol, node_text, ParamInfo};

pub fn extract_symbols_csharp(
    root: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Vec<NativeParsedSymbol> {
    let mut symbols = Vec::new();
    let mut stack = vec![root];

    while let Some(node) = stack.pop() {
        match node.kind() {
            "namespace_declaration" => {
                if let Some(symbol) = process_namespace_declaration(node, source, repo_id, rel_path)
                {
                    symbols.push(symbol);
                }
            }
            "class_declaration" => {
                if let Some(symbol) = process_type_like(node, source, repo_id, rel_path, "class") {
                    symbols.push(symbol);
                }
            }
            "interface_declaration" => {
                if let Some(symbol) =
                    process_type_like(node, source, repo_id, rel_path, "interface")
                {
                    symbols.push(symbol);
                }
            }
            "struct_declaration" | "record_declaration" => {
                if let Some(symbol) = process_type_like(node, source, repo_id, rel_path, "class") {
                    symbols.push(symbol);
                }
            }
            "enum_declaration" => {
                if let Some(symbol) = process_type_like(node, source, repo_id, rel_path, "type") {
                    symbols.push(symbol);
                }
            }
            "method_declaration" => {
                if let Some(symbol) = process_method_declaration(node, source, repo_id, rel_path) {
                    symbols.push(symbol);
                }
            }
            "constructor_declaration" => {
                if let Some(symbol) =
                    process_constructor_declaration(node, source, repo_id, rel_path)
                {
                    symbols.push(symbol);
                }
            }
            "property_declaration" => {
                if let Some(symbol) = process_variable_like(node, source, repo_id, rel_path) {
                    symbols.push(symbol);
                }
            }
            _ => {}
        }

        let child_count = node.child_count();
        for i in (0..child_count).rev() {
            if let Some(child) = node.child(i) {
                stack.push(child);
            }
        }
    }

    symbols
}

fn process_namespace_declaration(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Option<NativeParsedSymbol> {
    let name_node = node
        .child_by_field_name("name")
        .or_else(|| find_child_node(node, "qualified_name"))
        .or_else(|| find_child_node(node, "identifier"))?;

    let name = node_text(name_node, source).to_string();
    if name.is_empty() {
        return None;
    }

    // TS csharp.ts does NOT set visibility on namespace/module symbols.
    // Pass empty string which serializes as omitted (visibility is skipped
    // when empty by the NativeParsedSymbol serializer).
    let mut symbol = make_symbol(
        &name,
        "module",
        node,
        source,
        repo_id,
        rel_path,
        &[],
        None,
        &[],
        "",
        &extract_decorators(node, source),
    );
    symbol.exported = true;
    Some(symbol)
}

fn process_type_like(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
    kind: &str,
) -> Option<NativeParsedSymbol> {
    let name = extract_identifier(node, source)?;
    let generics = extract_type_parameters(node, source);
    let visibility = extract_visibility(node, source);
    let decorators = extract_decorators(node, source);

    // TS csharp.ts emits signature for class/interface/struct/record but NOT
    // for enum (which maps to kind "type" here). Force signature only for
    // the class-like kinds.
    let mut symbol = if kind == "type" {
        make_symbol(
            &name,
            kind,
            node,
            source,
            repo_id,
            rel_path,
            &[],
            None,
            &generics,
            &visibility,
            &decorators,
        )
    } else {
        super::common::make_symbol_with_forced_signature(
            &name,
            kind,
            node,
            source,
            repo_id,
            rel_path,
            &[],
            None,
            &generics,
            &visibility,
            &decorators,
        )
    };
    symbol.exported = visibility == "public";
    Some(symbol)
}

fn process_method_declaration(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Option<NativeParsedSymbol> {
    let name = extract_identifier(node, source)?;
    let visibility = extract_visibility(node, source);
    let modifiers = extract_modifiers(node, source);
    let params = extract_parameters(node, source);
    let mut returns = extract_return_type(node, source);

    if returns.is_none() {
        returns = if modifiers.iter().any(|m| m == "async") {
            Some("Task".to_string())
        } else {
            Some("void".to_string())
        };
    }

    let mut symbol = make_symbol(
        &name,
        "method",
        node,
        source,
        repo_id,
        rel_path,
        &params,
        returns.as_deref(),
        &[],
        &visibility,
        &extract_decorators(node, source),
    );
    symbol.exported = visibility == "public";
    Some(symbol)
}

fn process_constructor_declaration(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Option<NativeParsedSymbol> {
    let visibility = extract_visibility(node, source);
    let params = extract_parameters(node, source);

    // TS always emits signature for constructor; force it here.
    let mut symbol = super::common::make_symbol_with_forced_signature(
        "constructor",
        "constructor",
        node,
        source,
        repo_id,
        rel_path,
        &params,
        None,
        &[],
        &visibility,
        &extract_decorators(node, source),
    );
    symbol.exported = visibility == "public";
    Some(symbol)
}

fn process_variable_like(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Option<NativeParsedSymbol> {
    let name = extract_identifier(node, source)?;
    let visibility = extract_visibility(node, source);

    let mut symbol = make_symbol(
        &name,
        "variable",
        node,
        source,
        repo_id,
        rel_path,
        &[],
        None,
        &[],
        &visibility,
        &extract_decorators(node, source),
    );
    symbol.exported = visibility == "public";
    Some(symbol)
}

fn extract_identifier(node: Node<'_>, source: &[u8]) -> Option<String> {
    if node.kind() == "identifier" {
        let text = node_text(node, source).to_string();
        if !text.is_empty() {
            return Some(text);
        }
    }

    let mut stack = vec![node];
    while let Some(current) = stack.pop() {
        if current.kind() == "identifier" {
            let text = node_text(current, source).to_string();
            if !text.is_empty() {
                return Some(text);
            }
        }

        let child_count = current.child_count();
        for i in (0..child_count).rev() {
            if let Some(child) = current.child(i) {
                stack.push(child);
            }
        }
    }

    None
}

fn extract_visibility(node: Node<'_>, source: &[u8]) -> String {
    let Some(modifiers) = find_child_node(node, "modifiers") else {
        return String::new();
    };

    let mut cursor = modifiers.walk();
    for child in modifiers.children(&mut cursor) {
        if child.kind() == "accessibility_modifier" {
            let text = node_text(child, source);
            if matches!(text, "public" | "private" | "protected" | "internal") {
                return text.to_string();
            }
        }
    }

    String::new()
}

fn extract_modifiers(node: Node<'_>, source: &[u8]) -> Vec<String> {
    let Some(modifiers) = find_child_node(node, "modifiers") else {
        return Vec::new();
    };

    let mut out = Vec::new();
    let mut cursor = modifiers.walk();
    for child in modifiers.children(&mut cursor) {
        let text = node_text(child, source);
        if !text.is_empty() {
            out.push(text.to_string());
        }
    }
    out
}

fn extract_type_parameters(node: Node<'_>, source: &[u8]) -> Vec<String> {
    let Some(type_params) = find_child_node(node, "type_parameter_list") else {
        return Vec::new();
    };

    let mut generics = Vec::new();
    let mut cursor = type_params.walk();
    for child in type_params.children(&mut cursor) {
        if child.kind() == "type_parameter" {
            let text = node_text(child, source).to_string();
            if !text.is_empty() {
                generics.push(text);
            }
        }
    }
    generics
}

fn extract_parameters(node: Node<'_>, source: &[u8]) -> Vec<ParamInfo> {
    let Some(param_list) = find_child_node(node, "parameter_list") else {
        return Vec::new();
    };

    let mut params = Vec::new();
    let mut cursor = param_list.walk();
    for child in param_list.children(&mut cursor) {
        if child.kind() != "parameter" {
            continue;
        }

        let name = extract_identifier(child, source);
        if let Some(name) = name {
            let type_annotation = find_child_by_kind(child, "type", source).and_then(|t| {
                if t.is_empty() {
                    None
                } else {
                    Some(t)
                }
            });

            params.push(ParamInfo {
                name,
                type_annotation,
            });
        }
    }

    params
}

fn extract_return_type(node: Node<'_>, source: &[u8]) -> Option<String> {
    find_child_by_kind(node, "type", source)
        .and_then(|ret| if ret.is_empty() { None } else { Some(ret) })
}

fn extract_decorators(node: Node<'_>, source: &[u8]) -> Vec<String> {
    let mut decorators = Vec::new();
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        if child.kind() == "attribute_list" {
            decorators.push(node_text(child, source).to_string());
        }
    }
    decorators
}
